pageserver/src/tenant/timeline.rs

```diff
@@ -3377,27 +3406,38 @@ impl Timeline {
     ///
     /// At each iteration pop the top of the fringe (the layer with the highest Lsn)
     /// and get all the required reconstruct data from the layer in one go.
+    ///
+    /// Returns the completed keyspace and the keyspaces with image coverage. The caller
+    /// decides how to deal with these two keyspaces.
     async fn get_vectored_reconstruct_data_timeline(
         timeline: &Timeline,
         keyspace: KeySpace,
         mut cont_lsn: Lsn,
         reconstruct_state: &mut ValuesReconstructState,
         cancel: &CancellationToken,
         ctx: &RequestContext,
-    ) -> Result<KeySpace, GetVectoredError> {
+    ) -> Result<GetVectoredReconstructDataResult, GetVectoredError> {
         let mut unmapped_keyspace = keyspace.clone();
         let mut fringe = LayerFringe::new();
 
         let mut completed_keyspace = KeySpace::default();
+        let mut image_covered_keyspace = KeySpaceRandomAccum::new();
 
         loop {
             if cancel.is_cancelled() {
                 return Err(GetVectoredError::Cancelled);
             }
 
-            let keys_done_last_step = reconstruct_state.consume_done_keys();
+            let (keys_done_last_step, image_coverage_last_step) =
+                reconstruct_state.consume_done_keys();
             unmapped_keyspace.remove_overlapping_with(&keys_done_last_step);
+            unmapped_keyspace.remove_overlapping_with(&image_coverage_last_step);
             completed_keyspace.merge(&keys_done_last_step);
+            if !image_coverage_last_step.is_empty() {
+                for range in image_coverage_last_step.ranges {
```

Comment:
`image_coverage_last_step` doesn't have to be a keyspace. A `Range` is enough since we may only visit one image layer at a time.

Reply:
yep that makes sense, I'll change the `self.image_keys_done` to be a Option<Range<Key>>